packages/mui-system/src/cssVars/useCurrentColorScheme.ts

```diff
@@ -243,21 +258,6 @@ export default function useCurrentColorScheme<SupportedColorScheme extends strin
     return () => media.removeListener(handler);
   }, []);
 
-  // Save mode, lightColorScheme & darkColorScheme to localStorage
-  React.useEffect(() => {
```

Comment:
The intention of this was to save data to local storage on the client side when the application renders.

However, in multiple sessions, it causes the state to be updated multiple times.

Delete this logic and move the initial local storage setup to `React.useState()`

Reply:
Note: The local storage update logic is already in the `setMode` function.

Reply:
I would assume it should be the other way around. You should only update the local storage in the `useEffect` not in the `setState` methods, as it is basically a side effect, no?

Reply:
Updating local storage in this case is an action effect that should be done in the event handler so that it is triggered once when the mode transitions to another value.

As pointed out in https://youtu.be/HPoC-k7Rxwo?t=720 and the [React beta doc](https://beta.reactjs.org/learn/keeping-components-pure#where-you-can-cause-side-effects), the `useEffect` should be the last resort to run side effects.

Reply:
True, but you will need to manually ensure that every time you invoke `setState` you will update the `localStorage` with setting the appropriate key/values, depending on what changes. Having one `useEffect` to do this is more error prone in my opinion. Happy to try this and see how it goes, but in my opinion it unnecessarely complicates things. Maybe we can instead have a custom `setStateWithEffects` or something similar that will call the `setState` and update the local storage, and we should always invoke that function instead of directly `setState`. What do you think of this?

Reply:
The difference is that updating the localStorage in `useEffect` can have one flaw. If someone tries to read the `mode` in the localStorage in the next render before the Effect runs, it will return the previous one.

```md
// Initially, mode is `light`

// trigger setMode('dark') => React update the state and rerender

// if someone try to read the `mode` from localStorage
// the `mode` will be `light` instead of `dark`.

// an effect that set localStorage mode to `dark`
```

With this, I think it is safer to just save to localStorage at the same time as setMode since it is the only public API to set mode.

Reply:
Fair enough, and `setColorScheme` also makes sure we update the local storage too. As a last comment can we wrap all `localStora.setItem` with try and catch?

Reply:
> Fair enough, and setColorScheme also makes sure we update the local storage too

Correct.

> As a last comment can we wrap all localStora.setItem with try and catch?

This [PR](https://github.com/mui/material-ui/pull/34027) adds the `try {} catch (){}`. I'll update the PR when it is merged.

Reply:
Done!